### PR TITLE
#15859: Swap Integration using 0x API

### PR DESCRIPTION
**Pull request by onyb**
|GitHub|brave/brave-browser#15859|
-|-

### ⏮ Prior work

Please check the following PRs (contains demo videos): https://github.com/brave/ethereum-remote-client/pull/237, https://github.com/brave/ethereum-remote-client/pull/240, https://github.com/brave/ethereum-remote-client/pull/241, https://github.com/brave/ethereum-remote-client/pull/242

### 🔖 Summary (new in this PR)

* Support Ropsten network assets.
* **Implement transaction wait screen for token approval.**
* Priority sorted tokens list, aka poor man's substitute for Search.
  * Important assets are listed first, and the rest are sorted alphabetically.
* Improve accuracy of gas estimates.
* Lots of cleanups.
* Unit-tests (coverage report below).
    Screenshot 2021-06-09 at 21 15 30

#### Waiting screen for approve() transaction

⚠️ The first Failed approval was manually triggered by me and is part of the demo. Also, some parts of the video have been trimmed to keep it short.

https://user-images.githubusercontent.com/3684187/121385711-a4d2b500-c949-11eb-8e17-8a355071667f.mp4

#### On-chain transactions

* [LINK `approve()` transaction on mainnet](https://etherscan.io/tx/0xa7b5d321b68f94eb65e8f0cff362bc40ca5dc550b986e00b1ba1b6c49e806674).
* [ETH-DAI swap on ropsten](https://ropsten.etherscan.io/tx/0xc621eb03948301ca48c589c8a642b1371e21387f71c204cddaf645ea0e3ae885).
* [DAI `approve()` transaction on ropsten](https://ropsten.etherscan.io/tx/0xd5b0bf49e97f3f73d108268a9cc4a02558897d74ede2c1594b7e34d48fc40501).
* [DAI-ETH swap on ropsten](https://ropsten.etherscan.io/tx/0x079f8ab1ccbd3ffa3246154aff32a2cc38cd3efde69e33b3da3111d1e2092bcf).

####  Important notes for the QA

* If you feel lost, check the video recordings from the past PRs.
* Please note that the test-tokens on Ropsten used by 0x API may be different from the standard ones, hence they may not show up under the "Assets" section of the home page. Please use EtherScan for verifying balances if that's the case. This should not happen on the mainnet.
* Currently, the only swap pairs available on Ropsten for Swap are ETH and DAI. If you need a particular token for your tests, please let me know.
* If you're testing on the mainnet, please check if the rates are competitive, to avoid losing real money when swapping exotic tokens.
* Configuration related to the Swap feature can be found in the [`swap.config.js`](https://github.com/brave/ethereum-remote-client/blob/debug/swaps/ui/app/pages/swap/swap.config.js) file.

#### What to test?

(AAUIC = As a user I can)

* AAUIC swap from ETH to an ERC-20 token.
* AAUIC swap from an ERC-20 token to another ERC-20 token.
* AAUIC swap from an ERC-20 token to ETH.
* AAUIC wrap ETH.
	* This is an ETH-WETH swap, but it interacts directly with the `WETH9` contract and not with 0x exchange contracts.
* AAUIC unwrap WETH.
	* Same concept as wrapping ETH. 0x exchange contracts are not involved here.
* AAUIC swap between assets that may not have enough liquidity.
* AAUIC see gas-limit and gas-price estimations for swap transactions and be able to customize them.
* AAUIC see live quotes refreshed periodically.
* AAUIC set approval allowance before swapping an ERC-20 token. Check [this](http://help.matcha.xyz/en/articles/4285134-why-do-i-need-to-approve-my-tokens-before-i-can-trade-them) resource for more info.
	* Should allow setting of both fixed and unlimited allowances. Of course, you should test the former first - once an unlimited allowance is set, you wouldn't be given an option to change the allowance again.
* AAUIC see transaction parameters (to, from, amount, fees, data, etc) before signing the transaction.
* AAUIC easily go back to the Swap page after an ERC-20 token approval.
* AAUIC perform the entire flow using a hardware wallet.
* Brave is able to receive the Swap fees at its designated address. Please refer to the config file for details.

### What's next?

* [x] Hook into `/price` endpoint of 0x instead of `/quote`, which is faster to query and contains less data (hence, lighter on Redux).
* [x] Add slippage controls.
* [ ] (optional) Search assets.

### CODE REVIEW

**Comment by Genysys** [comment]
Looks good! I have a comment with regards to the approve. Usually , there isnt different approve function. 

The Dapp checks to see if an approval is require, then prompts for a user to approve, and then execute the transaction right after.

**Comment by onyb** [comment]
> I have a comment with regards to the approve. Usually , there isn't different approve function.
>
> The Dapp checks to see if an approval is require, then prompts for a user to approve, and then execute the transaction right after.

Technically speaking, there are always two transactions to sign => call to `approve()` and the main swap. Of course, the UX can make it seem that both these transactions happen back-to-back.

In our case, I decided to take the user back to the Swap page with all the previous data prefilled, so that they have an opportunity to revise the swap parameters (gas, swap amount, etc) and refresh the live quotes if they want. Matcha uses this kind of flow IIRC.

> I also think that we might need to either allow users to add their assets or search as opposed to hard coded asset lists.

I agree - it's a pain without the Search functionality. For the time being, the list is alphabetically sorted with the important assets listed first. I wanted to have an MVP out as soon as possible, so we can kickstart the review/QA process, while I continue with the polishes, and nice-to-have features.

I'm not sure about giving users the ability to add their own tokens. My two main reasons are:
* Users blind-sign the token address, even when using a hardware wallet, i.e., there's no way to verify what the assets being swapped are while signing the transaction. One of the reasons I decided to keep the assets list static was to prevent the possibility of introducing rogue tokens that disguise themselves with the same tickers.
* It seems overkill considering that this Swap feature may only remain live for a quarter or two.

**Comment by Genysys** [comment]
Sounds reasonable. LGTM!!

**Review by Genysys — APPROVED** [review]

File: ui/app/store/actions.js
Comment (diracdeltas) [inline ui/app/store/actions.js:704]: is it expected for allowance to be empty here?

Comment (diracdeltas) [reply ui/app/store/actions.js:704]: just wondering if it should throw an error instead of being set to maxAllowance in that case

Comment (diracdeltas) [reply ui/app/store/actions.js:704]: i saw the note in https://github.com/brave/ethereum-remote-client/pull/242 about the max allowance. my preference would be to error if no value is set. however i defer to you on whether this would be too much of a usability hurdle since i'm not sure what users expect.

Comment (onyb) [reply ui/app/store/actions.js:704]: After exploring different dapps, I think we should either set the allowance to the value of the swap or the available balance. The max allowance must be explicitly chosen if the user so wishes.

I'm going to push a fix for the default allowance.

**Comment by onyb** [comment]
**Update:** Screenshot with slippage controls and warning about setting approval allowances.

122866639-93fe4800-d328-11eb-9c1c-6d7910cbc369

File: ui/app/store/actions.js
Comment (onyb) [reply ui/app/store/actions.js]: > is it expected for allowance to be empty here?
> just wondering if it should throw an error instead of being set to maxAllowance in that case

Sorry, I forgot to respond to the above - an empty allowance means the user chose the value proposed as default. The `allowance` value is not empty only if the user provided a custom value.

Also, I've pushed the following fixes:
- use the swap `amount` as the default proposed limit, with a fallback to the token balance.
- added a warning tooltip, nudging the user to set approval allowance to 0 first. See discussion [here](https://github.com/brave/security/issues/480#issuecomment-865581044).

**Review by Genysys — APPROVED** [review]

**Comment by onyb** [comment]
**Update:** Added a privacy policy notice as per internal discussions on Slack.

|Privacy policy notice in Swap| Full text in "Settings > Security & Privacy" |
-|-
|Screenshot 2021-06-25 at 00 04 28|Screenshot 2021-06-25 at 00 04 46|

**Review by bbondy — APPROVED** [review]

**Comment by bbondy** [comment]
Merging to get some builds on the staging server. I will not deploy to production until there is QA and security sign off though.

**Pull request merged**
